File: src/epsilon_greedy.rs
use crate::trainer::print_python_like;
use rand::{rngs::SmallRng, Rng, SeedableRng};
use tch::Tensor;
pub enum EpsilonUpdateStrategy {
    AdaptativeEpsilon {
        min_epsilon: f32,
        max_epsilon: f32,
        min_reward: f32,
        max_reward: f32,
        eps_range: f32,
    },
    EpsilonCustomDecreasing {
        final_epsilon: f32,
        epsilon_decay: Box<dyn Fn(f32) -> f32 + Send + Sync>,
    },
    EpsilonLinearTrainingDecreasing {
        start: f32,
        end: f32,
        end_fraction: f32,
    },
    None,
}

impl EpsilonUpdateStrategy {
    fn update(
        &mut self,
        current_epsilon: f32,
        current_training_progress: f32,
        epi_reward: f32,
    ) -> f32 {
        match self {
            EpsilonUpdateStrategy::AdaptativeEpsilon {
                min_epsilon,
                max_epsilon,
                min_reward,
                max_reward,
                eps_range,
            } => {
                if epi_reward < *min_reward {
                    *max_epsilon
                } else {
                    let reward_range = *max_reward - *min_reward;
                    let min_update = *eps_range / reward_range;
                    let new_eps = (*max_reward - epi_reward) * min_update;
                    if new_eps < *min_epsilon {
                        *min_epsilon
                    } else {
                        new_eps
                    }
                }
            }
            EpsilonUpdateStrategy::EpsilonCustomDecreasing {
                final_epsilon,
                epsilon_decay,
            } => {
                let new_epsilon: f32 = (epsilon_decay)(current_epsilon);

                if *final_epsilon > new_epsilon {
                    current_epsilon
                } else {
                    new_epsilon
                }
            }
            EpsilonUpdateStrategy::EpsilonLinearTrainingDecreasing {
                start,
                end,
                end_fraction,
            } => {
                if current_training_progress > *end_fraction {
                    *end
                } else {
                    *start + current_training_progress * (*end - *start) / *end_fraction
                }
            }
            EpsilonUpdateStrategy::None => current_epsilon,
        }
    }
}

pub struct EpsilonGreedy {
    initial_epsilon: f32,
    current_epsilon: f32,
    rng: SmallRng,
    update_strategy: EpsilonUpdateStrategy,
}

impl Default for EpsilonGreedy {
    fn default() -> Self {
        Self::new(0.1, 42, EpsilonUpdateStrategy::None)
    }
}

impl EpsilonGreedy {
    pub fn new(epsilon: f32, seed: u64, update_strategy: EpsilonUpdateStrategy) -> Self {
        Self {
            initial_epsilon: epsilon,
            current_epsilon: epsilon,
            rng: SmallRng::seed_from_u64(seed),
            update_strategy,
        }
    }

    pub fn should_explore(&mut self) -> bool {
        let rng = self.rng.gen_range(0.0..1.0);
        // println!("{rng}");
        let r = self.current_epsilon != 0.0 && rng <= self.current_epsilon;
        // println!("{r}");
        r
    }

    pub fn get_action(&mut self, values: &Tensor) -> usize {
        // print_python_like(values);
        if self.should_explore() {
            // println!("{:?}", values.size()[0]);
            self.rng.gen_range(0..values.size()[0] as usize)
        } else {
            let a: i32 = values.argmax(0, true).try_into().unwrap();
            a as usize
        }
    }

    pub fn get_epsilon(&self) -> f32 {
        self.current_epsilon
    }

    pub fn reset(&mut self) {
        self.current_epsilon = self.initial_epsilon;
    }

    pub fn update(&mut self, current_training_progress: f32, epi_reward: f32) {
        self.current_epsilon = self.update_strategy.update(
            self.current_epsilon,
            current_training_progress,
            epi_reward,
        );
    }
}

File: python-src/epsilon_greedy.py
from small_rng_rs import SmallRng
import torch
from typing import Callable, Union
from dataclasses import dataclass


class EpsilonUpdateStrategy:
    @dataclass
    class AdaptativeEpsilon:
        min_epsilon: float
        max_epsilon: float
        min_reward: float
        max_reward: float
        eps_range: float

    @dataclass
    class EpsilonCustomDecreasing:
        final_epsilon: float
        epsilon_decay: Callable[[float], float]

    @dataclass
    class EpsilonLinearTrainingDecreasing:
        start: float
        end: float
        end_fraction: float

    def update(
        strategy,
        current_epsilon: float,
        current_training_progress: float,
        epi_reward: float,
    ) -> float:
        if isinstance(strategy, EpsilonUpdateStrategy.AdaptativeEpsilon):
            if epi_reward < strategy.min_reward:
                return strategy.max_epsilon
            else:
                reward_range = strategy.max_reward - strategy.min_reward
                min_update = strategy.eps_range / reward_range
                new_eps = (strategy.max_reward - epi_reward) * min_update
                return max(new_eps, strategy.min_epsilon)
        elif isinstance(strategy, EpsilonUpdateStrategy.EpsilonCustomDecreasing):
            new_epsilon = strategy.epsilon_decay(current_epsilon)
            return max(new_epsilon, strategy.final_epsilon)
        elif isinstance(
            strategy, EpsilonUpdateStrategy.EpsilonLinearTrainingDecreasing
        ):
            if current_training_progress > strategy.end_fraction:
                return strategy.end
            else:
                return (
                    strategy.start
                    + current_training_progress
                    * (strategy.end - strategy.start)
                    / strategy.end_fraction
                )
        else:
            return current_epsilon


class EpsilonGreedy:
    def __init__(
        self,
        epsilon: float,
        seed: int,
        update_strategy: Union[
            None,
            EpsilonUpdateStrategy.AdaptativeEpsilon,
            EpsilonUpdateStrategy.EpsilonCustomDecreasing,
            EpsilonUpdateStrategy.EpsilonLinearTrainingDecreasing,
        ],
    ):
        self.initial_epsilon = epsilon
        self.current_epsilon = epsilon
        self.rng = SmallRng(seed)
        self.update_strategy = update_strategy

    @classmethod
    def default(cls):
        return cls(0.1, 42, None)

    def should_explore(self) -> bool:
        rng = self.rng.uniform(0.0, 1.0, 1)[0]
        # print(round(rng, 7))
        r = self.current_epsilon != 0.0 and rng <= self.current_epsilon
        # print(r)
        return r

    def get_action(self, values: torch.Tensor) -> int:
        # print(values.to("cpu"))
        if self.should_explore():
            # print(values.size(0))
            v = self.rng.integer(0, values.size(0))
            return v
        else:
            return int(torch.argmax(values, keepdim=True).item())

    def get_epsilon(self) -> float:
        return self.current_epsilon

    def reset(self):
        self.current_epsilon = self.initial_epsilon

    def update(self, current_training_progress: float, epi_reward: float):
        self.current_epsilon = EpsilonUpdateStrategy.update(
            self.update_strategy,
            self.current_epsilon,
            current_training_progress,
            epi_reward,
        )
